[JSC] Add LLDB debugging infrastructure for WebAssembly (single VM)

hyjorc1 opened the pull request:
#### e03c10225cc8081620073a3eb8c551160d057fc7
<pre>
[JSC] Add LLDB debugging infrastructure for WebAssembly (single VM)
<a href="https://bugs.webkit.org/show_bug.cgi?id=299214">https://bugs.webkit.org/show_bug.cgi?id=299214</a>
<a href="https://rdar.apple.com/160971785">rdar://160971785</a>

Reviewed by Keith Miller.

This patch introduces a WebAssembly debugging solution that enables LLDB debugging
of WebAssembly code running in JavaScriptCore&apos;s IPInt tier through the GDB Remote
Serial Protocol. Currently supports single VM debugging scenarios.

Key components:
- WasmDebugServer: TCP server implementing GDB Remote Protocol
- Virtual address space management for LLDB compatibility
- Breakpoint and stepping support for WebAssembly bytecode
- Memory inspection and call stack visualization
- Comprehensive test framework with parallel execution

For detailed architecture and usage information, see:
1. Source/JavaScriptCore/wasm/debugger/README.md
2. JSTests/wasm/debugger/README.md

Canonical link: <a href="https://commits.webkit.org/300619@main">https://commits.webkit.org/300619@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e90afecbfe89701087dc613e5c5f11bc3792b7c0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/123175 "15 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/42889 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/33586 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/129882 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/75287 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/58b30c86-176e-4841-98dd-d8117e6f7fd4) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/125052 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/43613 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/51484 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/93635 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/62130 "Passed tests") | [❌ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/cdf8ee22-93a1-4281-85bb-21e35cf72648) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/126126 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/34759 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/110235 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/74264 "Found 1 new API test failure: TestWTF:WTF_Condition.OneProducerOneConsumerOneSlotTimeout (failure)") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/21352b2e-d049-448f-b800-297a718f71ae) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/33727 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/28391 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/73397 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/115334 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/104471 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/28617 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/132597 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/121706 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/50125 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/38164 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/102127 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/50501 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/106456 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/101983 "Found 1 new API test failure: WebKitGTK/TestInspector:/webkit/WebKitWebInspector/manual-attach-detach (failure)") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25942 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/47353 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/25559 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/46899 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/49980 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/55741 "Found 2 new failures in wasm/debugger/WasmDebugServerUtilities.cpp") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/151971 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/49450 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/38859 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/52800 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/51129 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

kmiller68 commented on Source/JavaScriptCore/jsc.cpp:
```diff
@@ -4016,6 +4017,7 @@ static void runInteractive(GlobalObject* globalObject)
 #endif
     fprintf(stderr, "  --destroy-vm               Destroy VM before exiting\n");
     fprintf(stderr, "  --can-block-is-false       Make main thread's Atomics.wait throw\n");
+    SAFE_FPRINTF(stderr, "  --wasm-debug[=port]        Enable WebAssembly debugging server (default port 1234)\n");
```
Why not just `fprintf` like the rest of the options?

hyjorc1 replied:
Changed to `fprintf` for now.

kmiller68 commented on Source/JavaScriptCore/runtime/WeakGCMap.h:
```diff
@@ -106,6 +106,11 @@ class WeakGCMap final : public WeakGCHashTable {
     void forEach(Func);
 
+    template<typename Func>
+    void forEach(Func) const;
```
Can be `const Invocable<IterationStatus(const Key&, const Value&)> auto&`

hyjorc1 replied:
Removed. We don't need that change now.

kmiller68 commented on Source/JavaScriptCore/runtime/WeakGCMapInlines.h:
```diff
@@ -88,4 +88,17 @@ inline void WeakGCMap<KeyArg, ValueArg, HashArg, KeyTraitsArg>::forEach(Func fun
 
+template<typename KeyArg, typename ValueArg, typename HashArg, typename KeyTraitsArg>
+template<typename Func>
+inline void WeakGCMap<KeyArg, ValueArg, HashArg, KeyTraitsArg>::forEach(Func func) const
```
Ditto.

hyjorc1 replied:
Removed. We don't need that change now.

kmiller68 commented on Source/JavaScriptCore/wasm/WasmFunctionParser.h:
```diff
@@ -444,6 +449,18 @@ auto FunctionParser<Context>::parse() -> Result
+                    localTypes->append(signature.argumentType(i));
+            }
+        }
+    }
```
I think you could push this into `m_context.addArguments` for IPIntGenerator.

hyjorc1 replied:
Fixed

kmiller68 commented on Source/JavaScriptCore/wasm/WasmFunctionParser.h:
```diff
@@ -463,6 +480,15 @@ auto FunctionParser<Context>::parse() -> Result
+                        localTypes->append(typeOfLocal);
+                }
+            }
+        }
```
Ditto for addLocal below. If you do this I don't think the parser needs to know about `m_debugInfo` at all.

hyjorc1 replied:
Fixed

kmiller68 commented on Source/JavaScriptCore/wasm/WasmIPIntGenerator.cpp:
```diff
@@ -540,9 +552,62 @@ class IPIntGenerator {
+                return extOp == static_cast<uint32_t>(ExtGCOpType::BrOnCast) || extOp == static_cast<uint32_t>(ExtGCOpType::BrOnCastFail);
+            }
+            return false;
+        }
```
This feels like the kind of thing that should be in WasmOps.h

hyjorc1 replied:
Fixed

kmiller68 commented on Source/JavaScriptCore/wasm/WasmIPIntGenerator.cpp:
```diff
@@ -651,15 +717,20 @@ class IPIntGenerator {
     bool m_usesRethrow { false };
     bool m_usesSIMD { false };
+
+    size_t m_functionStart { 0 };
```
Can we call this `m_functionStartByteOffset`

hyjorc1 replied:
Fixed

kmiller68 reviewed (commented):
Some comments

justinmichaud commented on Source/JavaScriptCore/wasm/debugger/README.md:
```diff
@@ -0,0 +1,227 @@
+
+### External IDE Debugging Support (WebContent)
+
+- **Status**: High Priority for Adobe Swift-to-WASM debugging
```
Maybe this isn't super clear for folks outside Apple

hyjorc1 replied:
Oops, removed.

justinmichaud commented on Source/JavaScriptCore/wasm/debugger/README.md:
````diff
@@ -0,0 +1,227 @@
+- **Goal**: Enable debugging WASM using external IDEs (VS Code, etc.) via LLDB
+- **Architecture**:
+  ```txt
+  External IDE ←→ LLDB ←→ RWI Bridge App ←→ webinspectord ←→ WebContent
````
This is Remote Web Inspector right?

hyjorc1 replied:
Yes, RWI stands for Remote Web Inspector.

JDevlieghere commented on Source/JavaScriptCore/wasm/debugger/README.md:
```diff
@@ -0,0 +1,226 @@
+- **View call stacks** across WebAssembly function calls
+- **Disassemble WebAssembly bytecode** in real-time
+
+The implementation follows the **GDB Remote Serial Protocol** standard, making it compatible with LLDB without requiring custom extensions.
```
This isn't entirely true, we do require a handful of extensions to the protocol: https://lldb.llvm.org/resources/lldbgdbremote.html#wasm-packets. It might be worth adding a link here.

hyjorc1 replied:
Oops, sorry that one was outdated. Fixed.

JDevlieghere commented on Source/JavaScriptCore/wasm/debugger/README.md:
````diff
@@ -0,0 +1,226 @@
+Terminal 2 - Connect LLDB:
+
+```bash
+lldb -o 'log enable gdb-remote packets' -o 'gdb-remote localhost:1234'
````
For now you still have to explicitly specify the `wasm` plugin. If you use the `gdb-remote` alias, lldb won't pick the correct process plugin and things like variables won't work.

```suggestion
lldb -o 'log enable gdb-remote packets' -o 'process connect --plugin wasm connect://localhost:1234'
```

hyjorc1 replied:
Thanks, that's right. Fixed.

kmiller68 commented on Source/JavaScriptCore/wasm/WasmIPIntSlowPaths.cpp:
This should probably be guarded above and below with `RELEASE_ASSERT`s that the debugger is enabled. That prevents attackers from using this to bootstrap control flow with a R/W.

hyjorc1 replied:
Updated.

kmiller68 commented on Source/WTF/wtf/Lock.h:
```diff
@@ -227,6 +227,9 @@ class WTF_CAPABILITY_SCOPED_LOCK Locker<T, std::enable_if_t<std::is_same_v<T, Lo
     template<typename>
     friend class DropLockForScope;
 
+    // Support Condition class to access private lock/unlock methods
+    friend class Condition;
+
```
Is this needed? I think this works today?

hyjorc1 replied:
Yeah, otherwise the build complains:
```
Condition.h:189:18: error: 'unlock' is a private member of 'WTF::Locker<WTF::Lock>' ...
```

kmiller68 commented on Tools/TestWebKitAPI/Tests/JavaScriptCore/glib/TestJSC.cpp:
```diff
@@ -4563,7 +4563,7 @@ static void testsJSCOptions()
     }, &optionsCount);
     g_assert_cmpuint(optionsCount, >, 100);
-    g_assert_cmpuint(optionsCount, <, 500);
+    g_assert_cmpuint(optionsCount, <, 510);
```
This seems like a bad check? What's it actually validating...

hyjorc1 replied:
Removed.

kmiller68 reviewed (approved):
r=me with some comments. I mostly just skimmed the code in debugger but it seems reasonable.

hyjorc1 committed:
[JSC] Add LLDB debugging infrastructure for WebAssembly (single VM)
https://bugs.webkit.org/show_bug.cgi?id=299214
rdar://160971785

Reviewed by Keith Miller.

This patch introduces a WebAssembly debugging solution that enables LLDB debugging
of WebAssembly code running in JavaScriptCore's IPInt tier through the GDB Remote
Serial Protocol. Currently supports single VM debugging scenarios.

Key components:
- WasmDebugServer: TCP server implementing GDB Remote Protocol
- Virtual address space management for LLDB compatibility
- Breakpoint and stepping support for WebAssembly bytecode
- Memory inspection and call stack visualization
- Comprehensive test framework with parallel execution

For detailed architecture and usage information, see:
1. Source/JavaScriptCore/wasm/debugger/README.md
2. JSTests/wasm/debugger/README.md

Canonical link: https://commits.webkit.org/300619@main
